plenum/server/consensus/view_change_service.py

```diff
@@ -212,14 +244,167 @@ def _validate(self, msg: Union[ViewChange, ViewChangeAck, NewView], frm: str) ->
+        # TODO: Actually we'll need to retrieve preprepares by ID from somewhere
+        self._data.preprepared = batches
+
+        # We finished a view change!
         self._data.waiting_for_new_view = False
+
+    def _clear_old_batches(self, batches: Dict[int, Any]):
+        for pp_seq_no in list(batches.keys()):
+            if pp_seq_no <= self._data.stable_checkpoint:
+                del batches[pp_seq_no]
+
+    @staticmethod
+    def _batch_id(batch: PrePrepare):
+        return BatchID(batch.viewNo, batch.ppSeqNo, batch.digest)
+
+    def _calc_checkpoint(self, vcs: List[ViewChange]) -> Optional[Checkpoint]:
+        checkpoints = []
+        for cur_vc in vcs:
+            for cur_cp in cur_vc.checkpoints:
+                # Don't add checkpoint to pretending ones if it is already there
+                if cur_cp in checkpoints:
+                    continue
+
+                # Don't add checkpoint to pretending ones if too many nodes already stabilized it
+                # TODO: Should we take into account view_no as well?
+                stable_checkpoint_not_higher = [vc for vc in vcs if cur_cp.seqNoEnd >= vc.stableCheckpoint]
+                if not self._data.quorums.strong.is_reached(len(stable_checkpoint_not_higher)):
+                    continue
+
+                # Don't add checkpoint to pretending ones if not enough nodes have it
+                have_checkpoint = [vc for vc in vcs if cur_cp in vc.checkpoints]
+                if not self._data.quorums.weak.is_reached(len(have_checkpoint)):
+                    continue
+
+                # All checks passed, this is a valid candidate checkpoint
+                checkpoints.append(cur_cp)
+
+        highest_cp = None
+        for cp in checkpoints:
+            # TODO: Should we take into account view_no as well?
+            if highest_cp is None or cp.seqNoEnd > highest_cp.seqNoEnd:
+                highest_cp = cp
+
+        return highest_cp
+
+    def _calc_batches(self, cp: Checkpoint, vcs: List[ViewChange]) -> List[BatchID]:
+        # TODO: Optimize this
+        batches = set()
+        for vc in vcs:
+            for _bid in vc.prepared:
+                bid = BatchID(*_bid)
+                if bid in batches:
+                    continue
+                if self._is_batch_prepared(bid, cp, vcs):
+                    batches.add(bid)
+
+            for _bid in vc.preprepared:
+                bid = BatchID(*_bid)
+                if bid in batches:
+                    continue
+                if self._is_batch_preprepared(bid, cp, vcs):
+                    batches.add(bid)
+
+        return sorted(batches)
+
+    def _is_batch_prepared(self, bid: BatchID, cp: Checkpoint, vcs: List[ViewChange]) -> bool:
+        if not self._is_inside_watermarks(bid, cp):
+            return False
+
+        def check(vc: ViewChange):
+            if bid.pp_seq_no <= vc.stableCheckpoint:
+                return False
+
+            for _some_bid in vc.prepared:
```

Comment:
This is the condition for ALL (n, d', v') from m': .... 
So, if it doesn't hold for at least one `_some_bid `, we should return.
Taking this, I think we need to have the following loop:
```
for _some_bid in vc.prepared:
    some_bid = BatchID(*_some_bid)
    if some_bid.pp_seq_no != bid.pp_seq_no:
        continue
    # not ( (v' < v) OR (v'==v and d'==d) )
    if some_bid.view_no > bid.view_no 
        return False
    if some_bid.view_no >= bid.view_no and some_bid.pp_digest != bid.pp_digest:
        return False
return True
```

Reply:
I've looked through code and original formula and now I remember why I implemented it that way. Formula indeed says `for all (n, d', v') from m'.P`, however the trick here is that there couldn't be two different records with same `n` in `P`, so I chose simpler implementation. I agree that it wasn't obvious at all and I should have added comment regarding it. Also, I think we need to add this (and other) invariant checkers to tests.

Reply:
No, this was not a simpler implementation, but a bit wrong one. 
There is a difference between "for all" and "at least one" predicates. 
The main difference is that "for all" is true even if there is no elements in a set, but "at least one" (as it was implemented) is false in this case.
And yes, I've already added tests for NewView builder (PR is coming soon)